### ISSUE TITLE
Consider using JSON-LD for the modules

### ISSUE THREAD

**Issue by csarven** [p2pcommons/specs#50]
p2pcommons modules are in plain JSON where some of the keys from the schemas appear to be borrowed from dat, whereas the p2pcommons keys are namespaced to `p2pcommons`.

Having the p2pcommons' schemas defined in its own JSON-LD will address extensibility, remixing and reuse. All that's needed is to write the schema in JSON-LD so that it can be referred to it via JSON-LD's `@context`. See example from the JSON-LD spec: https://www.w3.org/TR/json-ld/#the-context (or advanced examples under https://www.w3.org/TR/json-ld/#advanced-context-usage ).

This is essentially a matter of turning the definitions documented in https://github.com/p2pcommons/specs/blob/main/module.md into its own resource for p2pcommons schema.

The p2pcommons schema can also reuse terms from existing and well-established vocabularies! There is an ocean of them at your disposal. See for example: https://lov.linkeddata.es/dataset/lov/ . Applications like Hypergraph can also pick and choose to its own needs. See https://csarven.ca/linked-research-decentralised-web#vocabularies for a summary of some vocabularies relevant to research communication.

Both the application database and module graphs can be serialised in JSON-LD. It is relatively a minor change that shouldn't require change in application logic. Needless to say, test, test, test...

Having p2pcommons schema in JSON-LD is a *simple* point of interop with "Linked Data" / Web / Solid.. :)

**Comment by csarven** [comment]
See also `@context` examples from Activity Streams 2.0: https://www.w3.org/TR/activitystreams-core/

Context is in this resource:

`curl -H'Accept: application/ld+json' https://www.w3.org/ns/activitystreams`

`Content-Location` is https://www.w3.org/ns/activitystreams.jsonld

**Comment by chartgerink** [comment]
Do I understand correctly it would mean the metadata would look something like this at the end (i.e., adding the @context line):

```
{
  "title": "Test",
  "description": "Will I like JSON-LD",
  "url": "hyper://<hash>",
  "links": {
    "license": [
      {
        "href": "https://creativecommons.org/publicdomain/zero/1.0/legalcode"
      }
    ],
    "spec": [
      {
        "href": "https://p2pcommons.com/specs/module/1.0.0"
      }
    ]
  },
  "p2pcommons": {
    "@context": "https://p2pcommons.com/specs/module/1.0.0/spec.jsonld",
    "type": "profile",
    "subtype": "",
    "main": "",
    "avatar": "",
    "follows": [
      "<hash>"
    ],
    "contents": [
    ]
  }
}
```

where `https://p2pcommons.com/specs/module/1.0.0/spec.jsonld` refers to something along the lines of (but needs to be made specific wrt links)?

```
{
   "@context":
   {
      "type": "http://xmlns.com/foaf/0.1/Person",
      "subtype": "http://www.w3.org/2001/XMLSchema#",
      "main": "http://xmlns.com/foaf/0.1/name",
      "avatar": "http://xmlns.com/foaf/0.1/nick",
      "follows": "http://schema.org/affiliation",
      "contents": "http://schema.org/something"
}
```

**Comment by stain** [comment]
@chartgerink I think that's a good start!  I tried it in https://json-ld.org/playground/ but to make it come through to processing I also need to map the top-level `p2pcommons` and move the `@context` up there (or at least a context that maps `p2pcommons`.

The top-level properties seem like they would map quite straight forward to http://schema.org/CreativeWork, e.g. `title` is http://schema.org/name. I would rather use http://schema.org/Person than FOAF although that would not have social aspects like `follows` (`foaf:knows` perhaps).

You can certainly use other URIs like `hyper://` although of course their resolution is trickier than http/https.

```json
{
"@context":
   {
      "type": "http://xmlns.com/foaf/0.1/Person",
      "subtype": "http://www.w3.org/2001/XMLSchema#",
      "main": "http://xmlns.com/foaf/0.1/name",
      "avatar": "http://xmlns.com/foaf/0.1/nick",
      "follows": "http://schema.org/affiliation",
      "contents": "http://schema.org/something",
      "p2pcommons": "http://example.com/p2pcommons"
  },
  "title": "Test",
  "description": "Will I like JSON-LD",
  "url": "hyper://<hash>",
  "links": {
    "license": [
      {
        "href": "https://creativecommons.org/publicdomain/zero/1.0/legalcode"
      }
    ],
    "spec": [
      {
        "href": "https://p2pcommons.com/specs/module/1.0.0"
      }
    ]
  },
  "p2pcommons": {   
    "type": "profile",
    "subtype": "",
    "main": "",
    "avatar": "",
    "follows": [
      "<hash>"
    ],
    "contents": [
    ]
  }
}
```

**Comment by chartgerink** [comment]
Thanks @stain for thinking along with me on this 😊 

@csarven is this going in the direction you were thinking in your OP?

**Comment by almereyda** [comment]
There are two things to consider here. The first is serialising the a whole object as JSON-LD, and the second providing a JSON-LD context for existing JSON data.

In the JSON-LD specification [this is called *Interpreting JSON as JSON-LD*](https://w3c.github.io/json-ld-syntax/#interpreting-json-as-json-ld), which means the context "only" provides additional metadata for the keys used in an existing JSON schema, which will be later expanded into a whole JSON-LD document.

Given your context is

```
{
  "@context": {
    "type": "http://xmlns.com/foaf/0.1/Person",
    "subtype": "http://www.w3.org/2001/XMLSchema#",
    "main": "http://xmlns.com/foaf/0.1/name",
    "avatar": "http://xmlns.com/foaf/0.1/nick",
    "follows": "http://schema.org/affiliation",
    "contents": "http://schema.org/something",
    "p2pcommons": "http://example.com/p2pcommons"
  }
}
```

and your document is

```
{
  "title": "Test",
  "description": "Will I like JSON-LD",
  "url": "hyper://<hash>",
  "links": {
    "license": [
      {
        "href": "https://creativecommons.org/publicdomain/zero/1.0/legalcode"
      }
    ],
    "spec": [
      {
        "href": "https://p2pcommons.com/specs/module/1.0.0"
      }
    ]
  },
  "p2pcommons": {
    "type": "profile",
    "subtype": "",
    "main": "",
    "avatar": "",
    "follows": [
      "<hash>"
    ],
    "contents": [
    ]
  }
}
```

then these two will be merged into

```
{
  "@context": {
    "type": "http://xmlns.com/foaf/0.1/Person",
    "subtype": "http://www.w3.org/2001/XMLSchema#",
    "main": "http://xmlns.com/foaf/0.1/name",
    "avatar": "http://xmlns.com/foaf/0.1/nick",
    "follows": "http://schema.org/affiliation",
    "contents": "http://schema.org/something",
    "p2pcommons": "http://example.com/p2pcommons"
  },
  "title": "Test",
  "description": "Will I like JSON-LD",
  "url": "hyper://<hash>",
  "links": {
    "license": [
      {
        "href": "https://creativecommons.org/publicdomain/zero/1.0/legalcode"
      }
    ],
    "spec": [
      {
        "href": "https://p2pcommons.com/specs/module/1.0.0"
      }
    ]
  },
  "p2pcommons": {
    "type": "profile",
    "subtype": "",
    "main": "",
    "avatar": "",
    "follows": [
      "<hash>"
    ],
    "contents": [
    ]
  }
}
```

which is known as [compacted document form](https://w3c.github.io/json-ld-syntax/#compacted-document-form). In case the `url` key references the URL of the document itself, it can also be called `@id` in JSON-LD terms.

Since it describes the whole document, the context contains all keys used in the schema, hence here also: `title`, `description`, `url`, `links`, `license`, `href` and `spec`.

All in all supporting JSON-LD means your data can be parsed more easily by the outside world, and you maintain a set of terms, a glossary of your schema, a vocabulary of your domain for them. Internally you can continue processing the very same JSON as before.

**Comment by chartgerink** [comment]
Thanks @almereyda, that's super helpful. 🥂 

I can see this being recommended in a later version, as it would also require matching the items to their relevant existing contexts and writing new ones where necessary.

**Comment by csarven** [comment]
@chartgerink , Yes, I believe @stain covered my OP well. And, @almereyda's explanation matches as well.

You can essentially use any JSON-LD form (expanded, compacted, flattened..) that you prefer or find suitable for your environment. The encodings are equivalent (isomorphic) as far as the underlying graph goes for your data - for the most part. Meaning data can be (re-)serialized in any *concrete RDF syntax* (JSON-LD, Turtle, TriG, and RDFa in HTML, SVG, MathML..) or others like N-Triples, RDF/XML.. depending on the environment the data may be used in.

The trivial change here makes the system more interoperable based on open standards. Have fun.
